Streaming compressor for cached values

printercu opened the pull request:
Fixes #33910 , cc @jeremy 

For now when cache entry is serialized, it creates temporary marshaled string which is compressed later: https://github.com/rails/rails/blob/e925cb4d856088a815bf4a0cf27518d01bb4029d/activesupport/lib/active_support/cache.rb#L798-L799

We have faced that supervisor was killing job workers due to high memory usage while they were about writing to cache large objects (up to 1gb raw but with high compression ratio). This results in `Zlib::DataError: data error`.

I see that Marshal provides streaming api and can write serialized data to IO object. However Zlib doesn't support streams and stream-compression should be implemented manually. This both should decrease overall memory usage by avoiding storing large intermediate representation.

printercu committed:
Streaming compressor for cached values

Fixes #33910

printercu commented on activesupport/lib/active_support/cache/streaming_compressor.rb:
```diff
@@ -0,0 +1,109 @@
+        end
+
+        def handle_thread_exception(e)
+          Rails.logger.error(e)
```
Please advise the most appropriate solution for default exception handler in threads.

matthewd commented:
This seems like a lot of code for an edge case: if the cached value is too big to fit twice, it's not _that_ far away from being too big to fit once... if it were a three-liner I'd feel differently, but for this much bulk I'm inclined to say that the cache isn't a good fit for this use case. (I'd be looking for options that are all-streaming instead, say.)

That said, part of what seems to be making this so bulky is the thread handling: why is any of that necessary?

printercu commented:
`Marshal.dump` and `.load` are blocking calls and does not support chunky processing like this:

```ruby
serializer = Marshal.dump_chunky(value)
result << inflate_chunk(serializer.next_chunk) while serializer.any?
```

So it's necessary to put marshalling and compression on separate threads to make them both work simultaneously.

I understand that this streaming compressor handles only edge-case, maybe it's better to keep it in separate gem. However rails should provide API for switching compressor to make it work. This also brings ability to change compression method to another one (ex., zstd) without monkey-patching rails code.

Another option is to define `SerialCompressor` and make it default. Then add `compressor` option to read/fetch methods and use it only for large values: `cache.read(:key, compressor: :streaming)`.

I'll provide performance benchmarks soon, but I don't know how to make precise benchmarks of memory usage. In `top` streaming method shows ~250mb less memory usage than serial one when serializing `'a' * 300_000_000`.

printercu commented:
Added changes to address comments. Fixed tests and made some performance improvements, had to modify `dump, deflate, inflate`. Please see new versions.

I've prepared preview of custom compressors api in [this pr](https://github.com/printercu/rails/pull/1/files). WDYT?

Here is benchmark results (how much streaming serializer is slower than serial):

```
|--------------------------------------------|
|                | 2kb   | 200kb     | 200mb |
|--------------------------------------------|
| dump random    | 3.58x | same-ish  | 1.04x |
| dump repeated  | 7.07x | same-ish  | 1.08x |
|--------------------------------------------|
| load random    | 5.38x | 1.18x     | 1.01x |
| load repeated  | 8.40x | 1.55x     | 1.23x |
|--------------------------------------------|
```

<details>
 <summary>Benchmark source</summary>

```ruby
require 'securerandom'
require 'benchmark/ips'
require 'active_support/cache'

def bm(description, value)
  serial = ActiveSupport::Cache::SerialCompressor
  streaming = ActiveSupport::Cache::StreamingCompressor

  puts "#{description} dump (#{value.bytesize / 1.kilobyte}kb)"
  Benchmark.ips do |x|
    x.report('serial') { serial.dump(value) }
    x.report('streaming') { streaming.dump(value) }
    x.compare!
  end
  puts ''

  compressed = streaming.dump(value)
  puts "#{description} load (#{compressed.bytesize / 1.kilobyte}kb)"
  Benchmark.ips do |x|
    x.report('serial') { serial.load(compressed) }
    x.report('streaming') { streaming.load(compressed) }
    x.compare!
  end
  puts ''

  raise 'invalid' unless serial.load(compressed) == value
  raise 'invalid' unless streaming.load(compressed) == value
end

bm 'small random', SecureRandom.hex(1.kilobyte)
bm 'medium random', SecureRandom.hex(100.kilobyte)
bm 'large random', SecureRandom.hex(100.megabyte)

bm 'small repeated', 'a' * 2 * 1.kilobyte
bm 'medium repeated', 'a' * 2 * 100.kilobyte
bm 'large repeated', 'a' * 2 * 100.megabyte
```
</details>

schneems commented on activesupport/lib/active_support/cache/streaming_compressor.rb:
```diff
@@ -0,0 +1,110 @@
+          end
+        end
+        if compress_threshold
+          chunk = rio.read([DEFLATE_BUFFER_SIZE, compress_threshold].max)
```
Let's get rid of this extra array allocation with the `max` calculation

```ruby
size = DEFLATE_BUFFER_SIZE
size = compress_threshold if compress_threshold > size
chunk = rio.read([DEFLATE_BUFFER_SIZE, compress_threshold].max)
```

ohaddahan replied:
`compress_threshold && compress_threshold > DEFLATE_BUFFER_SIZE ? size = compress_threshold : size = DEFLATE_BUFFER_SIZE`
Removes the `if` and separate assigning to `size` in multiple lines.
Cleaner, personal opinion though.

printercu replied:
I'm affraid this may not work with current tests which has compress_threshold: 1. Inflate definitely does not work with  such small chunks. Will check deflate.

jonathanhefner replied:
Ruby 2.4 (min version for Rails 6) includes [an optimization](https://bugs.ruby-lang.org/issues/12172) which avoids allocating an extra array for certain invocations of `[].max`, which I think includes this case.

printercu replied:
@schneems rails master requires ruby >=2.4.1, which compiles this call to `opt_newarray_max 2`, instead of `newarray 2; 0015 opt_send_without_block <callinfo!mid:max...` in previous versions. Should I update the code?

schneems commented on activesupport/lib/active_support/cache/streaming_compressor.rb:
```diff
@@ -0,0 +1,110 @@
+        def binary_pipe
+          rio, wio = IO.pipe(binmode: true)
+          wio.set_encoding(Encoding::BINARY)
+          [rio, wio]
```
The return here requires an array allocation. Then ruby need another array to do this allocation 

```
rio, wio = binary_pipe
```

(when you assign this way, ruby allocates an array under the hood).

Can we restructure to not allocate those extra two arrays? One alternative could be to have binary_pipe yield those two values:

```ruby
require 'benchmark/ips'

def return_array
  a = "a"
  b = "b"
  return [a, b]
end

def use_block
  a = "a"
  b = "b"
  yield a, b
end

Benchmark.ips do |x|
  x.report("return array") { a, b = return_array }
  x.report("block") { 
    use_block do |a, b|
      # use a, b
    end
  }
  x.compare!
end
# Warming up --------------------------------------
#         return array   216.800k i/100ms
#                block   230.227k i/100ms
# Calculating -------------------------------------
#         return array      4.497M (± 2.1%) i/s -     22.547M in   5.015694s
#                block      5.261M (± 2.3%) i/s -     26.476M in   5.034858s

# Comparison:
#                block:  5261444.1 i/s
#         return array:  4497425.6 i/s - 1.17x  slower
```

printercu replied:
I'll change. It was using block version of IO.pipe before, but it had seldom strange errors "stream closed" in write. While block was finished after read is finished, which should occur after write finished. Race condition.

jonathanhefner replied:
I think `IO.pipe` will allocate an array for its return value regardless of whether you pass it a block (looking at the source [here](https://ruby-doc.org/core-2.5.0/IO.html#method-c-pipe)).  Therefore, to avoid an extra array allocation, it might be sufficient to do:

```ruby
def binary_pipe
  rw = IO.pipe(binmode: true)
  rw.last.set_encoding(Encoding::BINARY)
  rw
end
```

schneems reviewed (changes requested):
First of all, great job! The idea looks really exciting.

I did a minor perf review, there's two areas where we can decrease allocations right now. That could help speed up your benchmark numbers even more.


I also tried running my memory benchmarks against this change for CodeTriage, but apparently, all of my cache entries for my homepage are under the 1kb limit for compression so they didn't trigger your code.

I also tried to reproduce your results with your benchmark script and could not there is no `ActiveSupport::Cache::SerialCompressor` defined in your PR. Can you post a benchmark that I can run?

ohaddahan commented on activesupport/lib/active_support/cache/streaming_compressor.rb:
```diff
@@ -0,0 +1,110 @@
+
+      # Dumps value into string. Returns nil if `:compress_threshold` option
+      # is set and serialized value is less than given value.
+      def dump(value, **options)
```
Seems like the only use for `**options` is for `compress_threshold` , `**` operator is expensive.
Although it will not look as clean, explicitly passing `compress_threshold` will be quicker and given that it's only one argument now, seems reasonable.
https://gist.github.com/ohaddahan/f59aa5a261cb56e7b1ca96fb97a8e3fb
@schneems recently improved `try` by avoid unneeded `* splat` operator.

printercu replied:
I'll fix it. Thanks!

ohaddahan commented on activesupport/lib/active_support/cache/streaming_compressor.rb:
```diff
@@ -0,0 +1,110 @@
+        rio, wio = binary_pipe
+        within_safe_thread do
+          begin
+            Marshal.dump(value, wio)
```
`wio.write(value)` should be faster than `Marshal.dump(value, wio)`.
https://gist.github.com/ohaddahan/6946f93ae4f867d663529ed8fe4ace25

printercu replied:
This is key part of streaming compressor. Marshal.dump(value) creates string with whole serialized value. While Marshal.dump(value, io) dumps it to stream and paused, when stream is full, also discarding data which have already been read. Stream buffer size is defined by OS.

ohaddahan replied:
You're opening the `IO` on `#33` and ensuring closer on `#38`.
I don't see any logic change between `Marshal.dump(value, io)` vs. `wio.write(Mashsal.dump(value))`.
Can you please explain?

schneems replied:
@printercu This was confusing to me as well, `Marshal.dump` doesn't actually return a string object, but rather an IO object. I would still benchmark before making any kind of recommendation to change.

@ohaddahan you can benchmark this using the provided code in this comment (collapsed) https://github.com/rails/rails/pull/33920#issuecomment-423125837.

You can make a new rails example app from master https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#running-an-application-against-your-local-branch Then change it to point to a local copy or rails using `path:` in the gemfile. 

You can then get these changes in your local copy using 

```
git checkout -b printercu-streaming_compressor master
git pull https://github.com/printercu/rails.git streaming_compressor
```

From there you can make changes and benchmark yourself.  Actually looks like the branch is over here https://github.com/printercu/rails/pull/1/files.

ohaddahan replied:
@schneems @printercu not seeing a noticeable gain on medium + sizes.
https://gist.github.com/ohaddahan/ccf2abf42ef3034b13b3a5ee277ef0e0

BUT , when I tried benchmarking `wio.sync` I got consistent crashes. I'm not 100% sure this is thread safe.

printercu replied:
Looks like there is mistake in rubydoc:
```
[2] pry(main)> Marshal.dump(1).class
=> String < Object
[3] pry(main)> Marshal.dump(1, StringIO.new).class
=> StringIO < Data
```

ohaddahan replied:
@printercu you're correct, confirmed here.
https://gist.github.com/ohaddahan/c1bd9ec897b8675994f25a2101728e4e

ohaddahan replied:
`Marshal.dump` returns the `IO` it received.
`x = StringIO.new
y =  Marshal.dump(1, x)
x.object_id == y.object_id`

printercu commented:
@schneems can you please take SerialCompressor from https://github.com/printercu/rails/pull/1/files I've forgotten to mention that I run benchmarks in that branch

sham-sheer commented:
Whats up

sham-sheer commented:
hello
